python/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+fn pymicrovmi(_py: Python, m: &PyModule) -> PyResult<()> {
+    // init the env logger at module init
+    env_logger::init();
+
+    m.add_class::<Microvmi>()?;
+    m.add_class::<DriverType>()?;
+    m.add_class::<DriverInitParam>()?;
+
+    Ok(())
+}
+
+// fake enum impl for driver type
+#[pyclass]
+struct DriverType {}
+
+#[pymethods]
+impl DriverType {
+    #[classattr]
+    const HYPERV: u32 = 0;
+    #[classattr]
+    const KVM: u32 = 1;
+    #[classattr]
+    const VIRTUALBOX: u32 = 2;
+    #[classattr]
+    const XEN: u32 = 3;
+}
+
+// this will not be exported to Python
+#[derive(Debug, Copy, Clone)]
+enum DriverInitParamType {
+    KVMiUnixSocket = 0,
+}
+
+// exposing DriverInitParam to Python is bit more complicated
+// as it's an enum with associated data
+// the proposed implementation is exposing static method returning an
+// initialized instance of DriverInitParam struct
+/// Manages additional driver initialization parameters
+#[pyclass]
+#[derive(Debug, Clone)]
+struct DriverInitParam {
+    pub param_type: DriverInitParamType,
+    pub param_data_string: String,
+}
+
+#[pymethods]
+impl DriverInitParam {
+    /// initialize a DriverInitParam for the KVM driver, with a Unix socket
+    #[staticmethod]
+    fn kvmi_unix_socket(socket: &str) -> Self {
+        DriverInitParam {
+            param_type: DriverInitParamType::KVMiUnixSocket,
+            param_data_string: socket.to_string(),
+        }
+    }
+}
+
+/// Main class to interact with libmicrovmi
+// A class marked as unsendable will panic when accessed by another thread.
+// TODO: make Introspectable trait inherit Send trait, and make the drivers implementation
+// compatible
+#[pyclass(unsendable)]
+struct Microvmi {
+    driver: Box<dyn rapi::Introspectable>,
+}
+
+#[pymethods]
+impl Microvmi {
+    // enums are not available in PyO3 (yet)
+    // TODO: docstring is not exposed in Python (bug ?)
+    /// initializes libmicrovmi from the specified domain name
+    /// if driver_type is None, every driver compiled in libmicrovmi will be tested,
+    /// and the first one that succeeds will be returned, or an error
+    ///
+    /// Args:
+    ///     domain_name (str): the domain name
+    ///     driver_type (int, optional): the hypervisor driver type on which the library should be initialized.
+    ///     init_param (DriverInitParam, optional): additional initialization parameters for driver initialization
+    #[new]
+    #[args(domain_name, driver_type = "None", init_param = "None")]
+    fn new(
+        domain_name: &str,
+        driver_type: Option<u32>,
+        init_param: Option<DriverInitParam>,
+    ) -> PyResult<Self> {
+        info!("Microvmi Python init");
+        // convert Python DriverType to rust API DriverType
+        debug!(
+            "Microvmi Python init driver_type: {:?}, init_param: {:?}",
+            driver_type, init_param
+        );
+        let rust_driver_type: Option<rapi::DriverType> = if let Some(drv_type) = driver_type {
```

Comment:
IIRC using `map` on the option type would be better than using `if else`

Reply:
I tried, but had an issue with the `?` operator
~~~rust
        let rust_driver_type: Option<rapi::DriverType> = driver_type.map(|drv_type| match drv_type {
                DriverType::HYPERV => Ok(rapi::DriverType::HyperV),
                DriverType::KVM => Ok(rapi::DriverType::KVM),
                DriverType::VIRTUALBOX => Ok(rapi::DriverType::VirtualBox),
                DriverType::XEN => Ok(rapi::DriverType::Xen),
                _ => Err(PyValueError::new_err(format!(
                    "Invalid value for DriverType: {}",
                    drv_type
                ))),
            }?);
~~~

~~~
error[E0277]: the `?` operator can only be used in a closure that returns `Result` or `Option` (or another type that implements `Try`)
   --> python/src/lib.rs:106:85
    |
106 |            let rust_driver_type: Option<rapi::DriverType> = driver_type.map(|drv_type| match drv_type {
    |   __________________________________________________________________________-__________^
    |  |__________________________________________________________________________|
    | ||
107 | ||                 DriverType::HYPERV => Ok(rapi::DriverType::HyperV),
108 | ||                 DriverType::KVM => Ok(rapi::DriverType::KVM),
109 | ||                 DriverType::VIRTUALBOX => Ok(rapi::DriverType::VirtualBox),
...   ||
114 | ||                 ))),
115 | ||             }?);
    | ||              ^
    | ||______________|
    | |_______________cannot use the `?` operator in a closure that returns `microvmi::api::DriverType`
    |                 this function should return `Result` or `Option` to accept `?`
    |
    = help: the trait `Try` is not implemented for `microvmi::api::DriverType`
    = note: required by `from_error`
~~~

:(

Reply:
see my suggestion